plugins/crypto/mbedtls/ua_pki_mbedtls.c

```diff
@@ -236,251 +238,158 @@ reloadCertificates(CertInfo *ci) {
+            if(!x509_name_eq(&cert->issuer, &i->subject))
+                continue;
+            if(prev == NULL)
+                return i;
+            prev = NULL;
+        }
+        /* Switch to search in the ctx->skIssue list */
+        stack = (stack != &ci->certificateIssuerList) ? &ci->certificateIssuerList : NULL;
+    } while(stack);
+    return NULL;
+}
 
-    if(ci->trustListFolder.length == 0 &&
-       ci->issuerListFolder.length == 0 &&
-       ci->revocationListFolder.length == 0 &&
-       ci->certificateTrustList.raw.len == 0 &&
-       ci->certificateIssuerList.raw.len == 0 &&
-       ci->certificateRevocationList.raw.len == 0) {
-        UA_LOG_WARNING(UA_Log_Stdout, UA_LOGCATEGORY_SERVER,
-                       "PKI plugin unconfigured. Accepting the certificate.");
-        return UA_STATUSCODE_GOOD;
+static UA_Boolean
+mbedtlsCheckRevoked(CertInfo *ci, mbedtls_x509_crt *cert) {
+    for(mbedtls_x509_crl *crl = &ci->certificateRevocationList; crl; crl = crl->next) {
+        if(x509_name_eq(&cert->subject, &crl->issuer))
+            return (mbedtls_x509_crt_is_revoked(cert, crl) != 0);
     }
+    return false;
+}
 
-    /* Parse the certificate */
-    mbedtls_x509_crt remoteCertificate;
-
-    /* Temporary Object to parse the trustList */
-    mbedtls_x509_crt *tempCert = NULL;
-
-    /* Temporary Object to parse the revocationList */
-    mbedtls_x509_crl *tempCrl = NULL;
-
-    /* Temporary Object to identify the parent CA when there is no intermediate CA */
-    mbedtls_x509_crt *parentCert = NULL;
-
-    /* Temporary Object to identify the parent CA when there is intermediate CA */
-    mbedtls_x509_crt *parentCert_2 = NULL;
-
-    /* Flag value to identify if the issuer certificate is found */
-    int issuerKnown = 0;
-
-    /* Flag value to identify if the parent certificate found */
-    int parentFound = 0;
+#define UA_MBEDTLS_MAX_CHAIN_LENGTH 10
 
-    mbedtls_x509_crt_init(&remoteCertificate);
-    int mbedErr = mbedtls_x509_crt_parse(&remoteCertificate, certificate->data,
-                                         certificate->length);
-    if(mbedErr) {
-        /* char errBuff[300]; */
-        /* mbedtls_strerror(mbedErr, errBuff, 300); */
-        /* UA_LOG_WARNING(data->policyContext->securityPolicy->logger, UA_LOGCATEGORY_SECURITYPOLICY, */
-        /*                "Could not parse the remote certificate with error: %s", errBuff); */
-        return UA_STATUSCODE_BADSECURITYCHECKSFAILED;
+static UA_StatusCode
+mbedtls_verifyChain(CertInfo *ci, mbedtls_x509_crt *cert,
+                    mbedtls_x509_crt **old_issuers, int depth) {
+    /* Maxiumum chain length */
+    if(depth == UA_MBEDTLS_MAX_CHAIN_LENGTH)
+        return UA_STATUSCODE_BADCERTIFICATECHAININCOMPLETE;
+
+    /* Verification Step: Validity Period */
+    if(mbedtls_x509_time_is_future(&cert->valid_from) ||
+       mbedtls_x509_time_is_past(&cert->valid_to))
+        return UA_STATUSCODE_BADCERTIFICATETIMEINVALID;
```

Comment:
Isn't it required to check what certificate type we are actually looking at? For several error conditions there are versions for Application Instance Certificates (AICs) and for Issuer Certificates.

If we look at an AIC `UA_STATUSCODE_BADCERTIFICATETIMEINVALID` is ok here. But for an Issuer Certificate it should be `UA_STATUSCODE_BADCERTIFICATEISSUERTIMEINVALID` or not? This should apply to further conditions as well.

Reply:
We now distinguish between the two error codes.